docs/src/options.md
# Solver options

This documentation lists all solver options and the default value in `()`

In general these options can be set as follows if you're using JuMP:

```
m = Model(optimizer_with_attributes(CS.Optimizer, "option_name"=>option_value))
```

## `logging` (`[:Info, :Table]`)

Current possible symbols
- :Info
    - Shows info about how many variables are part of the model
    - Info about which constraints are present in the model
- :Table
    - Shows a table about the current solver status

Output will be something like
```
# Variables: 5
# Constraints: 2
 - # Inequality: 2

   #Open      #Closed         Incumbent             Best Bound        Time [s]  
================================================================================
     2           0                -                   44.20            0.0003  
```
  
## `table` (`TableSetup(...)`)

Defines the exact table setup. The actual default is:

```
TableSetup(
        [:open_nodes, :closed_nodes, :incumbent, :best_bound, :duration],
        ["#Open", "#Closed", "Incumbent", "Best Bound", "[s]"],
        [10,10,20,20,10]; 
        min_diff_duration=5.0
)
```

which means that the open/closed nodes, the incumbent and the best bound is shown besides the duration of the optimization process. 
- `[10,10,20,20,10]` gives the width of each column
- `min_diff_duration=5.0` => a new row is added every 5 seconds or if a new solution was found.

For satisfiability problems the incumbent and best bound are `0` so you could remove them. I'll probably add that ;)

## `time_limit` (`Inf`)

Time limit for backtracking in seconds. If reached before the problem was solved or infeasibility was proven will return the status `MOI.TIME_LIMIT`.

## `seed` (`1`)

Some parts of the constraint solver use random numbers. Nevertheless everything should be reproducable which is the default case. You can make it "truly" random by setting a random `seed`.

## `rtol` (`1e-6`)

Defines the relative tolerance of the solver.

## `atol` (`1e-6`)

Defines the absolute tolerance of the solver.

## `lp_optimizer` (`nothing`)

It is advised to use a linear problem solver like [Cbc.jl](https://github.com/JuliaOpt/Cbc.jl) if you have a lot of linear constraints and an optimization problem. The solver is used to compute bounds in the optimization steps.

## `traverse_strategy` (`:Auto`)

You can chose a traversal strategy for you model with this option. The default is choosing depending on the model. In feasibility problems depth first search is chosen and in optimization problems best first search.
Other options:
- `:BFS` => Best First Search
- `:DFS` => Depth First Search
- `:DBFS` => Depth First Search until solution was found then Best First Search

## `branch_strategy` (`:Auto`)

You can chose a branch strategy for you model with this option.

**Options:**
- `:IMPS` => Infeasible and Minimum Possibility Search
  - This is currently the automatic default
  - Chooses the next variable based on 
    - whether the variable is part of the objective function
    - an infeasibility counter for each variable
    - the number of open possibilities
- `:ABS` => [Activity Based Search](https://arxiv.org/pdf/1105.6314.pdf)
  - It is based on that paper but doesn't implement value selection. For further options see `activity`


## `activity` 

The following options can be set with `activity.` i.e `"activity.decay" => 0.9`. These options are only taken into consideration when the `branch_strategy` option is set to `:ABS`

### `decay` (0.999)

The activity of variables decays when they are not used in the current node. In the following it is written as $\gamma$.
$X$ are the variables and $X^{\prime}$ denotes variables that have been changed. 
$D(x)$ is the domain of the variable $x$ and 
$A(x)$ is the activity.

```math
\begin{aligned}
\forall x \in X& \text { s.t. } |D(x)| > 1 &: A(x) = A(x) \cdot \gamma \\
\forall x \in X^{\prime}& &: A(x)=A(x)+1
\end{aligned}
```

### `max_probes` (`10`)

When activity based search is selected the search space gets probed by using a random variable selection strategy to initialize the activity of each variable.

The probing can be stopped by either hitting `max_probes` or when one can be confident to a certain degree that the approximated activity is good enough. (see `max_confidence_deviation`)

### `max_confidence_deviation` (`20`)

Probing as explained in `max_probes` can be stopped when each variable has an approximated activity when is in a specified bound. The bound can be changed using this option.

More precisely probing is stopped when this range
```math
\left[\tilde{\mu_{A}}(x)-t_{0.05, n-1} \cdot \frac{\tilde{\sigma_{A}}(x)}{\sqrt{n}}, \tilde{\mu_{A}}(x)+t_{0.05, n-1} \cdot \frac{\tilde{\sigma_{A}}(x)}{\sqrt{n}}\right]
```

is within $\pm$ `max_confidence_deviation` % of the empirical mean.

## `branch_split` (`:Auto`)

You can define how the variable is split into two branches with this option. 
Normally the smallest value is chosen for satisfiability problems and depending on the coefficient of the variable in the objective either the smallest or biggest is chosen as a single choice for optimization problems.
Other options:
- `:Smallest` Smallest value on the left branch and rest on the right branch
- `:Biggest` same as smallest but splits into biggest value as a single choice and the rest the second choice
- `:InHalf` takes the mean value to split the problem into two branches of equal size

## `simplify` (`True`)

Defines whether the solver should spend some time and effort to simplify constraints.
Curently this works for the following:
- combining `!=` constraints to all different constraints
- combining `sum(x) == V` overlapping all different constraints to introduce more constraints
- combining `x .>= y` constraints to a `GeqSet` constraint which helps with bound computation

It can be turned off to not waste time if this isn't applicable or not wanted for whatever reason.

## `all_solutions` (`false`)

You can set this to `true` to get **all feasible** solutions. This can be used to get all solutions for a sudoku for example but maybe shouldn't be used for an optimization problem. Nevertheless I leave it here so you be able to use it even for optimization problems and get all feasible solutions.

Look at `all_optimal_solutions` if you only want all solutions with the same optimum.

## `all_optimal_solutions` (`false`)

You can set this to `true` to get **optimal solutions**. If you have a feasibility problem you can also use `all_solutions` but for optimization problems this will only return solutions with the same best incumbent.


## `backtrack` (`true`)

To solve the problem completely normally backtracking needs to be used but for some problems like certain sudokus this might not be necessary. This option is mostly there for debugging reasons to check the search space before backtracking starts.

## `max_bt_steps` (`typemax(Int)`)

You can set the maximum amount of backtracking steps with this option. Probably you only want to change this if you want to debug some stuff.

## `backtrack_sorting` (`true`)

If set to true the order of new nodes is determined by their best bound. Otherwise they will be traversed in order they were added to the stack.

## `keep_logs` (`false`)

Sometimes you might be interested in the exact way the problem got solved then you can set this option to `true` to get the full search tree.

To save the logs as a json file you need to run:

```
m = Model()
...
com = CS.get_inner_model(m)

CS.save_logs(com, "FILENAME.json")
```

Additionally because the mapping from JuMP can be different to your internal mapping you can use:

```
CS.save_logs(com, "FILENAME.json", :x => x)
```
if `x` is/are your variable/variables and if you have more variables:

```
CS.save_logs(com, "FILENAME.json", :x => x, :y => y)
```
etc...

## `solution_type` (`Float64`)

Defines the type of `best_bound` and `incumbent`. Normally you don't want to change this as JuMP only works with `Float` but if you work directly using MathOptInterface you can use this option.

CHANGELOG.md
# ConstrainSolver.jl - Changelog

## v0.6.3 (17th of January 2021)
- Use anti pruning in reified constraints 

## v0.6.2 (16th of January 2021)
- Bugfix when linear constraint has only variables with coefficient 0 like `x <= x` => `0x <= 0`

## v0.6.1 (15th of January 2021)
- Bugfix if binary variable is constrained directly in `@variable`
- Use `CS.get_inner_model` to get the `ConstraintSolverModel`
  - Usage of `JuMP.backend` as before will fail
- Support for strictly less than and greater than with `<` and `>`
- Refactoring
  - Using bridges for `>=` and `>` in indicator and reified constraint
  - Combining `==` and `<=` into `src/constraints/linear_constraints.jl`

## v0.6.0 (15th of December 2020) 
- **Dropped support for Julia v1.0 and v1.1**
- Implementation of Activity Based Search
  - `"branch_strategy" => :ABS`
- Throw error if there is a dimension mismatch in a `TableSet`
- Bugfix in `simplify!` check that one inner index isn't used more than once

## v0.5.3 (12th of December 2020)
- Bugfix for optimization with reified and indicator constraints
  - If inner constraint does not define bound variables

## v0.5.2 (11th of December 2020)
- Bugfix for `x == x` constraints
- Bugfix for directly infeasible constraint inside reified and indicator 

## v0.5.1 (10th of December 2020)
- Bugfix in indicator constraint 
  - If inner constraint is fixed and indicator tries to get active
    - Double check that the inner constraint is solved

## v0.5.0 (8th of December 2020)
- Using a priority queue for faster `get_next_node`
- Removed `further_pruning`

## v0.4.1 (8th of December 2020)
- Using faster version of strongly connected components with lower memory footprint
- Bugfix in reified constraint:
  - If inner constraint can't be activated it shouldn't be solved
    - Bug found by @hakank see [Issue #202](https://github.com/Wikunia/ConstraintSolver.jl/issues/202)

## v0.4.0 (29th of November 2020)
**Improvements for graph coloring**
- Use LP solver also for single variable objectives
- Combine several `x != y` constraints into an all different constraint
- Combine several `a >= x` constraints with the same `a` into a vector constraints
  - used for better bounds using all different constraints

## v0.3.1 (16th of November 2020)
- Added `copy` function for constraint structs for latest JuMP/MOI versions

## v0.3.0 (11th of July 2020)
- Reified constraint [#171](https://github.com/Wikunia/ConstraintSolver.jl/pull/171)

## v0.2.2 (26th of June 2020)
- Actually use best bound [#175](https://github.com/Wikunia/ConstraintSolver.jl/pull/175)
- Select next var based on objective (still hacky solution) [#176](https://github.com/Wikunia/ConstraintSolver.jl/issues/176)

## v0.2.1 (26th of June 2020)
- Bugfixes in indicator constraint [#170](https://github.com/Wikunia/ConstraintSolver.jl/issues/170)
  - Calling finished constraints and other functions for i.e `TableConstraint` as an inner constraint
  - Use correct best bound when inactive vs active

## v0.2.0 (17th of June 2020)
- Bugfix for indicator constraints
    - support for TableConstraint in Indicator

## v0.1.8 (15th of June 2020)
- Support for indicator constraints
    - i.e. `@constraint(m, b => { x + y <= 10 })`

## v0.1.7 (22th of May 2020)
- Better feasibility and pruning in `==`
- **Bugfixes:**
  - Correct set of change ptr in `EqualSet` for faster/correct pruning
  - Call to `isapprox_discrete` in `eq_sum`
  - Fixed threshold rounding

## v0.1.6 (11th of May 2020)
- Reduction of memory allocations in `TableConstraint`
- Pruning in `EqualSet`

## v0.1.5 (6th of May 2020)
- **Bugfixes:**
  - EqualSet feasibility: Check if other vars have value + no memory allocation
  - Call `call_finished_pruning!(com)` after second `prune!` before backtracking

## v0.1.4 (6th of May 2020)
- Added `is_constraint_solved` functions to check whether problem gets actually solved
- **Bugfixes:**
  - Correct incremental update in table constraint
  - Fixed `restore_pruning_constraint!` in table
  
## v0.1.3 (4th of May 2020)
- **Bugfixes:**
  - Use correct offset in table constraint `support` and `residues`

## v0.1.2 (27th of April 2020)
- Table constraint
  - `@constraint(m, x in CS.TableSet(2dArr))`

## v0.1.1 (15th of April 2020)
- CS.Integers
  - i.e `@variable(m, x, CS.Integers([1,2,5]))`
- **Bugfixes:**
  - return infeasible if start fixing is infeasible [#132](https://github.com/Wikunia/ConstraintSolver.jl/pull/132) 

## 0.1.0 (7th of April 2020)
Initial implementation of the solver:
Basic constraints: 
- AllDifferent
- Linear functions `==`, `>=`, `<=`, `!=`

src/options.jl
function get_auto_traverse_strategy(com::CS.CoM)
    return com.sense == MOI.FEASIBILITY_SENSE ? :DFS : :BFS
end

function get_traverse_strategy(com; options = SolverOptions())
    if options.traverse_strategy == :DBFS
        return isempty(com.solutions) ? Val(:DFS) : Val(:BFS)
    end
    return Val(options.traverse_strategy)
end

function get_auto_branch_strategy(com::CS.CoM)
    return :IMPS # Infeasible and Minimum Possiblity Search
end

function get_branch_strategy(; options = SolverOptions())
    strategy = options.branch_strategy
    return Val(strategy)
end

function get_branch_split(; options = SolverOptions())
    strategy = options.branch_split
    return Val(strategy)
end

const POSSIBLE_OPTIONS = Dict(
    :traverse_strategy => [:Auto, :BFS, :DFS, :DBFS],
    :branch_strategy => [:Auto, :ABS, :IMPS],
    :branch_split => [:Auto, :Smallest, :Biggest, :InHalf],
)

function SolverOptions()
    logging = [:Info, :Table]
    table = TableSetup(
        [
            CS.TableCol(:open_nodes, "#Open", Int, 10, :center),
            CS.TableCol(:closed_nodes, "#Closed", Int, 10, :center),
            CS.TableCol(:incumbent, "Incumbent", Float64, 20, :center),
            CS.TableCol(:best_bound, "Best Bound", Float64, 20, :center),
            CS.TableCol(:duration, "Time [s]", Float64, 10, :center),
        ],
        Dict(:min_diff_duration => 5.0),
    )
    seed = 1
    traverse_strategy = :Auto
    branch_strategy = :Auto
    branch_split = :Auto
    backtrack = true
    max_bt_steps = typemax(Int)
    backtrack_sorting = true
    keep_logs = false
    rtol = 1e-6
    atol = 1e-6
    solution_type = Float64
    all_solutions = false
    all_optimal_solutions = false
    lp_optimizer = nothing
    time_limit = Inf
    no_prune = false
    decay = 0.999
    max_probes = 10
    max_confidence_deviation = 20
    simplify = true

    return SolverOptions(
        logging,
        table,
        time_limit,
        seed,
        traverse_strategy,
        branch_strategy,
        branch_split,
        backtrack,
        max_bt_steps,
        backtrack_sorting,
        keep_logs,
        rtol,
        atol,
        solution_type,
        all_solutions,
        all_optimal_solutions,
        lp_optimizer,
        no_prune,
        ActivityOptions(decay, max_probes, max_confidence_deviation),
        simplify,
    )
end

function combine_options(options)
    defaults = SolverOptions()
    options_dict = Dict{Symbol,Any}()
    for kv in options
        if !in(kv[1], fieldnames(SolverOptions))
            @error "The option $(kv[1]) doesn't exist."
        else
            moi_key = MOI.RawParameter(string(kv[1]))
            if is_possible_option_value(moi_key, kv[2])
                options_dict[kv[1]] = kv[2]
            else
                @error "The option $(kv[1]) doesn't have $(kv[2]) as a possible value. Possible values are: $(POSSIBLE_OPTIONS[moi_key])"
            end
        end
    end

    for fname in fieldnames(SolverOptions)
        if haskey(options_dict, fname)
            setfield!(
                defaults,
                fname,
                convert(fieldtype(SolverOptions, fname), options_dict[fname]),
            )
        end
    end
    return defaults
end

function is_possible_option_value(option_param::MOI.RawParameter, value)
    option = Symbol(option_param.name)
    if haskey(POSSIBLE_OPTIONS, option)
        return value in POSSIBLE_OPTIONS[option]
    end
    return true
end
